presto-spi/src/main/java/com/facebook/presto/spi/relation/LogicalRowExpressions.java

```diff
@@ -384,14 +444,64 @@ private RowExpression notCallExpression(RowExpression argument)
         {
             return new CallExpression("not", functionResolution.notFunction(), BOOLEAN, singletonList(argument));
         }
+    }
 
-        private boolean isConjunctionOrDisjunction(RowExpression expression)
+    private class ConvertNormalFormVisitor
+            implements RowExpressionVisitor<RowExpression, Form>
+    {
+        @Override
+        public RowExpression visitSpecialForm(SpecialFormExpression specialForm, Form clauseJoiner)
         {
-            if (expression instanceof SpecialFormExpression) {
-                Form form = ((SpecialFormExpression) expression).getForm();
-                return form == AND || form == OR;
+            if (!isConjunctionOrDisjunction(specialForm)) {
+                return specialForm;
+            }
+
+            // normalize arguments
+            RowExpression left = specialForm.getArguments().get(0).accept(new ConvertNormalFormVisitor(), clauseJoiner);
```

Comment:
Are you referring to Lines 463-466? If so, it's because `clauseJoiner` is the highest separator (for CNF it's AND) and the left and right expressions represent 2 sets of valid clauses. So for ex. if we're doing CNFs then we have a set of ANDS for left and right, so `left AND right` is already in normal form. Same for DNF: `left OR right` is fine as left and right are already organized in sets of ors.